libraries/tock-register-interface/src/peripheral.rs

```diff
@@ -0,0 +1,288 @@
+/// ```ignore
+/// #[derive(Clone, Copy)]
+/// struct RealFoo<
+///     P: tock_registers::MmioPointer,
+///     B: tock_registers::BusAdapter = tock_registers::DirectAdapter>
+///     { ... }
+/// ```
+/// This argument is used to support unusual chip designs (mainly some LiteX
+/// designs). Because the bus adapter can change the size of registers (e.g. a
+/// register with a data type of `u8` may take up more than 1 byte in the MMIO
+/// address space), `peripheral!` allows you to specify that an offset is to be
+/// inferred:
+/// ```
+/// tock_registers::peripheral! {
+///     #[allow_bus_adapter]
+///     Foo {
+///         // The first register is still at offset 0.
+///         0x0 => a: u8 { Read },
+///
+///         // This register has an inferred offset, which may depend on the bus
+///         // adapter passed to RealFoo.
+///         _ => b: u32 { Write },
+///     }
+/// }
+/// ```
+/// Note that a padding field cannot be followed by a field with an inferred
+/// offset.
+///
+/// # Generated `macro_rules!` macro
+/// In addition to the `Foo` trait and `RealFoo` trait, `peripheral!` also
+/// generates a `macro_rules!` macro with the same name as the trait:
+/// ```ignore
+/// #[macro_export]
+/// macro_rules! Foo { ... }
+/// ```
+/// This is currently unused, but is emitted because we may want to use it in
+/// the future, and adding it later would be a breaking change.
+///
+/// # Unsafe operations
+/// Registers can be unsafe to use:
+/// ```
+/// tock_registers::peripheral! {
+///     Foo {
+///         // A write-only register that is unsafe to write.
+///         0x0 => a: u8 { UnsafeWrite },
+///
+///         // A register that can safely be read from or unsafely be written to
+///         0x1 => b: u8 { Read, UnsafeWrite },
+///
+///         // A register can also be unsafe to read (for use if reading it
+///         // triggers a memory-unsafe operation).
+///         0x2 => c: u8 { UnsafeRead },
+///
+///         // Any combination of readable, writable, safe, and unsafe is
+///         // possible.
+///     }
+/// }
+/// ```
+/// These registers implement the [`UnsafeRead`] and [`UnsafeWrite`] traits
+/// instead of their safe equivalents.
+///
+/// # `cfg` attributes
+/// Peripheral declarations can have `cfg` attributes:
+/// ```
+/// tock_registers::peripheral! {
+///     // This peripheral is only present if the has_foo feature is enabled.
+///     #[cfg(feature = "has_foo")]
+///     Foo {
+///         0x0 => ctrl: u8 { Write },
+///
+///         // This register is only present if the bounce feature is enabled.
+///         #[cfg(feature = "bounce")]
+///         0x1 => bounce: u8 { Read, Write },
+///
+///         // Often, you will need to place a register with an inferred offset
+///         // (or padding) after a register with a `cfg` attribute, because the
+///         // offset may change depending on the crate's enabled features.
+///         _ => next_register: u8 {},
```

Comment:
Upon more reflection, I'm starting to lean against allowing inferred offsets in any context. I think this is an area where explicit beats implicit. Inferred padding length is a reasonable ergonomic thing, but for any 'real' / accessible thing, I think we should require explicit placement / definition.

Reply:
> because the offset may change depending on the crate's enabled features.

Does this mean there is hardware where, say, the Control register is at address 0x8 but then on another variant of the chip that supports bounce the Control register would be at address 0xC? That's pretty scary.

It seems like this implicit placement would have to cascade to all future registers in the peripheral, so you could just use `_` as the address for all of them? That seems like something we don't want.